src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1610,6 +1611,28 @@ RangeTableListContainsOnlyReferenceTables(List *rangeTableList)
 				return false;
 			}
 		}
+		else if (rangeTableEntry->rtekind == RTE_FUNCTION)
+		{
+			List *functionList = rangeTableEntry->functions;
+
+			/*
+			 * We should allow both intermediate results function
+			 * and any other immutable functions.
+			 */
+			if ((list_length(functionList) == 1 &&
+				 ContainsReadIntermediateResultFunction(linitial(functionList))))
+			{
+				continue;
+			}
+			else if (!contain_mutable_functions((Node *) functionList))
+			{
+				continue;
+			}
+			else
+			{
+				return false;
+			}
+		}
```

Comment:
I wonder, could we call `HasRecurringTuples` on each RTE or combine those functions somehow?

Reply:
`HasRecurringTuples` goes recursively down into the node, which is not intended at this point. Because, we're already going deep into the query tree as needed with the one level above function.